zdm/analyze_cube.py

```diff
@@ -0,0 +1,477 @@
+""" Codes to analyze cube outputs """
+import os
+import numpy as np
+import pickle
+import glob
+
+import math
+import scipy
+
+import pandas
+
+from matplotlib import pyplot as plt
+
+from zdm import io
+from zdm import iteration
+
+from IPython import embed
+
+def slurp_cube(input_file:str, prefix:str, outfile:str, 
+               nsurveys, debug:bool=False):
+    """ Slurp the cube ASCII output files and write 
+    lC and ll into a numpy savez file
+
+    Args:
+        input_file (str): parameter file used to generate the cube
+        prefix (str): prefix on the output files
+        outfile (str): output file name.  Should have .npz extension
+        nsurveys (int): Number of surveys in the analysis. 
+        debug (int, optional): Debug?
+    """
+    # Grab em.  The order doesn't matter
+    files = glob.glob(prefix+'*.out') 
+
+    # Init
+    input_dict=io.process_jfile(input_file)
+    _, cube_dict, vparam_dict = iteration.parse_input_dict(input_dict)
+    PARAMS = list(vparam_dict.keys())
+
+    # Prep
+    order, iorder = iteration.set_orders(cube_dict['parameter_order'], PARAMS)
+    cube_shape = iteration.set_cube_shape(vparam_dict, order)
+
+    param_shape = np.array([0]+cube_shape)[iorder].tolist()[:-1]
+    ll_cube = np.zeros(param_shape)
+    lC_cube = np.zeros(param_shape)
+    ll_cube[:] = -9e9
+
+    survey_items = ['lls', 'DM_z', 'N', 'SNR', 'Nex']
+    names = ['icube'] + PARAMS
+    for ss in range (nsurveys):
+        names += [item+f'_{ss}' for item in survey_items]
+    names += ['ll']
+    
+    # Loop on cube output files
+    for dfile in files:
+        print(f"Loading: {dfile}")
+        df = pandas.read_csv(dfile, header=None, delimiter=r"\s+", names=names)
+
+        for index, row in df.iterrows():
+            # Unravel
+            r_current = np.array([0]+list(np.unravel_index(
+                        int(row.icube), cube_shape, order='F')))
+            current = r_current[iorder][:-1] # Truncate lC
+            # Ravel me back
+            idx = np.ravel_multi_index(current, ll_cube.shape)
+            # Set
+            ll_cube.flat[idx] = row.ll
+            lC_cube.flat[idx] = row.lC
+        # Check
+        if debug:
+            embed(header='69 of analyze')
+    
+    # Write
+    np.savez(outfile, ll=ll_cube, lC=lC_cube)
+    print(f"Wrote: {outfile}")
+
+
+def get_bayesian_data(lls:np.ndarray, 
```

Comment:
OK, I clearly need to extend this to 2D in the future.

Reply:
:+1: